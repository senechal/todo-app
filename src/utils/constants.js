export const INLINE_EDIT_FOCUS_DELAY = 0;
export const INLINE_EDIT_ENTER_KEY = 13;
export const EMPTY_STRING = '';
export const EMPTY_COUNTER = 0;
export const SINGLE_LEFT = 1;
export const INCREMENT = 1;

export const TEXT_ADD_ITEM = 'Add item';
export const TEXT_HEADER_MSG = 'What needs to be done?';
export const TEXT_DELETE_ITERM = 'Delete';
export const TEXT_EMPTY_LIST = 'Empty list.';
export const TEXT_ALL_DONE = 'All Done!';
export const TEXT_ONE_ITEM_LEFT = '1 item left.';
export const TEXT_ITEMS_LEFT = ' items left.';

export const REDUX_PERSISTENT_STATE = 'savedReducer';

export const TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT = 'inline-edit-item-content';
export const TEST_ID_INLINE_EDIT_INPUT = 'inline-edit-input';
export const TEST_ID_TODO_ITEM_CHECKBOX = 'todo-item-checkbox';
export const TEST_ID_TODO_ITEM_DELETE_BUTTON = 'todo-item-delete-button';
export const TEST_ID_TODO_LIST_ADD_BUTTON = 'todo-list-add-button';
export const TEST_ID_TODO_LIST_CHECKBOX = 'todo-list-checkbox';
