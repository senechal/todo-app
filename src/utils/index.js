import {
  EMPTY_COUNTER,
  TEXT_EMPTY_LIST,
  TEXT_ALL_DONE,
  TEXT_ONE_ITEM_LEFT,
  TEXT_ITEMS_LEFT,
  SINGLE_LEFT,
} from './constants';

export const getCounterText = (counter, all) => {
  if (counter <= EMPTY_COUNTER && all === EMPTY_COUNTER) {
    return TEXT_EMPTY_LIST;
  }
  if (counter === EMPTY_COUNTER && all > EMPTY_COUNTER) {
    return TEXT_ALL_DONE;
  }
  if (counter === SINGLE_LEFT) {
    return TEXT_ONE_ITEM_LEFT;
  }
  return `${counter}${TEXT_ITEMS_LEFT}`;
};

export const focusComponent = (element) => {
  element.current.focus();
};
export const blurComponent = (element) => {
  element.current.blur();
};
