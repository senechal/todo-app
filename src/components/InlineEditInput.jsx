import React, {
  useState, useRef, useEffect, useCallback,
} from 'react';
import PropTypes from 'prop-types';
import delay from 'lodash/delay';
import {
  INLINE_EDIT_FOCUS_DELAY,
  INLINE_EDIT_ENTER_KEY,
  EMPTY_STRING,
  TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT,
  TEST_ID_INLINE_EDIT_INPUT,
} from 'utils/constants';
import { focusComponent, blurComponent } from 'utils';
import { Input, ItemContent } from './styles';

const InlineEditInput = ({
  value, onStartEdit, onEndEdit, ...props
}) => {
  const inputRef = useRef();
  const [edit, setEdit] = useState(false);

  const handleEdit = useCallback(() => {
    setEdit(true);
    delay(() => {
      focusComponent(inputRef);
      onStartEdit();
    }, INLINE_EDIT_FOCUS_DELAY);
  }, [onStartEdit]);

  const handleBlur = useCallback(() => {
    setEdit(false);
    onEndEdit();
  }, [onEndEdit]);

  const handleKeyPress = ({ charCode }) => {
    if (charCode === INLINE_EDIT_ENTER_KEY) {
      blurComponent(inputRef);
    }
  };

  useEffect(() => {
    if (value === EMPTY_STRING && edit === false) {
      handleEdit();
    }
  }, [value, edit, handleEdit]);

  return (
    <>
      <Input
        data-testid={TEST_ID_INLINE_EDIT_INPUT}
        ref={inputRef}
        {...props}
        value={value}
        onBlur={handleBlur}
        hide={!edit}
        onKeyPress={handleKeyPress}
      />
      <ItemContent
        data-testid={TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT}
        onClick={handleEdit}
        hide={edit}
      >
        {value}
      </ItemContent>
    </>
  );
};

InlineEditInput.propTypes = {
  value: PropTypes.string,
  onStartEdit: PropTypes.func,
  onEndEdit: PropTypes.func,
};

InlineEditInput.defaultProps = {
  value: undefined,
  onStartEdit: Function.prototype,
  onEndEdit: Function.prototype,
};

export default InlineEditInput;
