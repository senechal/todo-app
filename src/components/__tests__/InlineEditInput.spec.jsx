import React from 'react';
import 'jest-styled-components';
import { render, fireEvent } from '@testing-library/react';
import { focusComponent, blurComponent } from 'utils';
import {
  TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT,
  TEST_ID_INLINE_EDIT_INPUT,
  INLINE_EDIT_ENTER_KEY,
} from 'utils/constants';
import InlineEditInput from '../InlineEditInput';

jest.mock('lodash/delay', () => (fn) => { fn(); });
jest.mock('../../utils', () => ({
  ...jest.requireActual('../../utils'),
  blurComponent: jest.fn(),
  focusComponent: jest.fn(),
}));

describe('Testing InlineEditInput', () => {
  const props = {
    value: 'value',
    onStartEdit: jest.fn(),
    onEndEdit: jest.fn(),
    onChange: jest.fn(),
  };

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render correctly without props', () => {
    const { container } = render(<InlineEditInput />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with props', () => {
    const { container } = render(<InlineEditInput {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly, focus on input and call onStartEdit after user click on content', () => {
    const { container, getByTestId } = render(<InlineEditInput {...props} />);

    const itemContainer = getByTestId(TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT);
    fireEvent.click(itemContainer);

    expect(props.onStartEdit).toBeCalled();
    expect(focusComponent).toBeCalled();
    expect(container).toMatchSnapshot();
  });

  it('should render correctly and call onEndEdit after user blur from input', () => {
    const { container, getByTestId } = render(<InlineEditInput {...props} />);

    const itemContainer = getByTestId(TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT);
    fireEvent.click(itemContainer);

    const input = getByTestId(TEST_ID_INLINE_EDIT_INPUT);
    fireEvent.blur(input);

    expect(props.onEndEdit).toBeCalled();
    expect(container).toMatchSnapshot();
  });

  it('should blur from input on press enter', () => {
    const { container, getByTestId } = render(<InlineEditInput {...props} />);

    const itemContainer = getByTestId(TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT);
    fireEvent.click(itemContainer);

    const input = getByTestId(TEST_ID_INLINE_EDIT_INPUT);
    fireEvent.keyPress(input, { charCode: 33 });

    expect(blurComponent).not.toBeCalled();

    fireEvent.keyPress(input, { charCode: INLINE_EDIT_ENTER_KEY });
    expect(blurComponent).toBeCalled();

    expect(container).toMatchSnapshot();
  });

  it('should render correctly, focus on input and call onStartEdit if value is empty without and not editing', () => {
    const { container, getByTestId } = render(<InlineEditInput {...{ ...props, value: '' }} />);
    const itemContainer = getByTestId(TEST_ID_INLINE_EDIT_INPUT_ITEM_CONTENT);
    fireEvent.click(itemContainer);

    expect(props.onStartEdit).toBeCalled();
    expect(focusComponent).toBeCalled();
    expect(container).toMatchSnapshot();
  });
});
