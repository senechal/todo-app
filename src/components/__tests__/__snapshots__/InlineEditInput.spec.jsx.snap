// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing InlineEditInput should blur from input on press enter 1`] = `
.c1 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

.c1:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c0 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

<div>
  <input
    class="c0"
    data-testid="inline-edit-input"
    type="text"
    value="value"
  />
  <div
    class="c1"
    data-testid="inline-edit-item-content"
  >
    value
  </div>
</div>
`;

exports[`Testing InlineEditInput should render correctly and call onEndEdit after user blur from input 1`] = `
.c1 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

.c1:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c0 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

<div>
  <input
    class="c0"
    data-testid="inline-edit-input"
    type="text"
    value="value"
  />
  <div
    class="c1"
    data-testid="inline-edit-item-content"
  >
    value
  </div>
</div>
`;

exports[`Testing InlineEditInput should render correctly with props 1`] = `
.c1 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

.c1:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c0 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

<div>
  <input
    class="c0"
    data-testid="inline-edit-input"
    type="text"
    value="value"
  />
  <div
    class="c1"
    data-testid="inline-edit-item-content"
  >
    value
  </div>
</div>
`;

exports[`Testing InlineEditInput should render correctly without props 1`] = `
.c1 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

.c1:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c0 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

<div>
  <input
    class="c0"
    data-testid="inline-edit-input"
    type="text"
    value=""
  />
  <div
    class="c1"
    data-testid="inline-edit-item-content"
  />
</div>
`;

exports[`Testing InlineEditInput should render correctly, focus on input and call onStartEdit after user click on content 1`] = `
.c1 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

.c1:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c0 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

<div>
  <input
    class="c0"
    data-testid="inline-edit-input"
    type="text"
    value="value"
  />
  <div
    class="c1"
    data-testid="inline-edit-item-content"
  >
    value
  </div>
</div>
`;

exports[`Testing InlineEditInput should render correctly, focus on input and call onStartEdit if value is empty without and not editing 1`] = `
.c1 {
  color: #4d4d4d font-size:16px;
  height: 20px;
  box-sizing: border-box;
  display: none;
  width: calc(100% - 62px);
}

.c1:hover {
  border-bottom: 1px solid #e6e6e6;
}

.c0 {
  color: #4d4d4d;
  outline: 0;
  border-width: 0 0 1px;
  border-color: #4d4d4d;
  font-size: 16px;
  height: 20px;
  box-sizing: border-box;
  width: calc(100% - 62px);
}

<div>
  <input
    class="c0"
    data-testid="inline-edit-input"
    type="text"
    value=""
  />
  <div
    class="c1"
    data-testid="inline-edit-item-content"
  />
</div>
`;
